Sources/AsyncHTTPClient/SSLContextCache.swift

```diff
@@ -0,0 +1,102 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Logging
+import NIO
+import NIOConcurrencyHelpers
+import NIOSSL
+
+class SSLContextCache {
+    private var state = State.activeNoThread
+    private let lock = Lock()
+    private var sslContextCache = LRUCache<BestEffortHashableTLSConfiguration, NIOSSLContext>()
+    private let threadPool = NIOThreadPool(numberOfThreads: 1)
+
+    enum State {
+        case activeNoThread
+        case active
+        case shutDown
+    }
+
+    init() {}
+
+    func shutdown() {
+        self.lock.withLock { () -> Void in
+            switch self.state {
+            case .activeNoThread:
+                self.state = .shutDown
+            case .active:
+                self.state = .shutDown
+                self.threadPool.shutdownGracefully { maybeError in
+                    precondition(maybeError == nil, "\(maybeError!)")
+                }
+            case .shutDown:
+                preconditionFailure("SSLContextCache shut down twice")
+            }
+        }
+    }
+
+    deinit {
+        assert(self.state == .shutDown)
+    }
+}
+
+extension SSLContextCache {
+    func sslContext(tlsConfiguration: TLSConfiguration,
+                    eventLoop: EventLoop,
+                    logger: Logger) -> EventLoopFuture<NIOSSLContext> {
+        let earlyExit = self.lock.withLock { () -> EventLoopFuture<NIOSSLContext>? in
+            switch self.state {
+            case .activeNoThread:
+                self.state = .active
+                self.threadPool.start()
+                return nil
+            case .active:
+                return nil
+            case .shutDown:
+                struct SSLContextCacheShutdownError: Error {}
+                return eventLoop.makeFailedFuture(SSLContextCacheShutdownError())
+            }
+        }
```

Comment:
Returning an optional future from this block seems odd to me. Do we maybe want to return an optional `Error` and use that to fail a future outside the lock?